ast/edit/match.go

```diff
@@ -0,0 +1,436 @@
+	}
+	return true
+}
+
+func matchLogicalOperator(o, oo ast.LogicalOperatorKind) bool {
+	if o >= 0 && o != oo {
+		return false
+	}
+	return true
+}
+
+func matchPackage(p *ast.Package, op *ast.Package) bool {
+	if !matchString(p.Path, op.Path) ||
+		!matchString(p.Package, op.Package) {
+		return false
+	}
+	if len(p.Files) > len(op.Files) {
+		return false
+	}
+	if len(p.Files) == 0 {
+		return true
+	}
+	var matched bool
+	for i := 0; i < len(op.Files) && !matched; i++ {
+		if len(p.Files) > len(op.Files)-i-1 {
+			// stop searching, the sub-sequence is longer than the rest
+			break
+		}
+		for ii := 0; i < len(p.Files) && !matched; ii++ {
+			if !match(p.Files[ii], op.Files[i+ii]) {
+				// match not found, continue searching
+				break
+			}
+			if ii == len(p.Files)-1 {
+				// match found
+				matched = true
+			}
+		}
+	}
+	return matched
+}
+
+func matchFile(f *ast.File, of *ast.File) bool {
+	if !matchString(f.Name, of.Name) {
+		return false
+	}
+	if !match(f.Package, of.Package) {
+		return false
+	}
+	if len(f.Imports) > len(of.Imports) {
+		return false
+	}
+	if len(f.Imports) == 0 {
+		return true
+	}
+	var matched bool
+	for i := 0; i < len(of.Imports) && !matched; i++ {
+		if len(f.Imports) > len(of.Imports)-i-1 {
+			break
+		}
+		for ii := 0; i < len(f.Imports) && !matched; ii++ {
+			if !match(f.Imports[ii], of.Imports[i+ii]) {
+				break
+			}
+			if ii == len(f.Imports)-1 {
+				matched = true
+			}
+		}
+	}
+	return matched
+}
+
+func matchBlock(b *ast.Block, ob *ast.Block) bool {
+	if len(b.Body) > len(ob.Body) {
+		return false
+	}
+	if len(b.Body) == 0 {
+		return true
+	}
+	var matched bool
+	for i := 0; i < len(ob.Body) && !matched; i++ {
+		if len(b.Body) > len(ob.Body)-i-1 {
+			break
+		}
+		for ii := 0; i < len(b.Body) && !matched; ii++ {
+			if !match(b.Body[ii], ob.Body[i+ii]) {
+				break
+			}
+			if ii == len(b.Body)-1 {
+				matched = true
+			}
+		}
+	}
+	return matched
+}
+
+func matchPackageClause(c *ast.PackageClause, oc *ast.PackageClause) bool {
+	return match(c.Name, oc.Name)
+}
+
+func matchImportDeclaration(d *ast.ImportDeclaration, od *ast.ImportDeclaration) bool {
+	return match(d.As, od.As) && match(d.Path, od.Path)
+}
+
+func matchOptionStatement(s *ast.OptionStatement, os *ast.OptionStatement) bool {
+	return match(s.Assignment, os.Assignment)
+}
+
+func matchExpressionStatement(s *ast.ExpressionStatement, os *ast.ExpressionStatement) bool {
+	return match(s.Expression, os.Expression)
+}
+
+func matchReturnStatement(s *ast.ReturnStatement, os *ast.ReturnStatement) bool {
+	return match(s.Argument, os.Argument)
+}
+
+func matchVariableAssignment(a *ast.VariableAssignment, oa *ast.VariableAssignment) bool {
+	return match(a.ID, oa.ID) && match(a.Init, oa.Init)
+}
+
+func matchMemberAssignment(a *ast.MemberAssignment, oa *ast.MemberAssignment) bool {
+	return match(a.Member, oa.Member) && match(a.Init, oa.Init)
+}
+
+func matchCallExpression(e *ast.CallExpression, oe *ast.CallExpression) bool {
```

Comment:
I like the convention of using `p` to indicate a pattern in this file, would be good to continue that here.

Reply:
It was more `statement`, `other-statement`. I could change everything to `p`attern, `n`ode.

Reply:
I like that.  One of the troubles with a tool like this is that the pattern and the thing you're matching are the same data type so it's easy to get confused.